tools/clang_tidy/lib/src/command.dart

```diff
@@ -0,0 +1,130 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert' show utf8, LineSplitter;
+import 'dart:io' as io;
+
+import 'package:path/path.dart' as path;
+import 'package:process_runner/process_runner.dart';
+
+/// The url prefix for issues that must be attached to the directive in files
+/// that disables linting.
+const String issueUrlPrefix = 'https://github.com/flutter/flutter/issues';
+
+/// Lint actions to apply to a file.
+enum LintAction {
+  /// Run the linter over the file.
+  lint,
+
+  /// Ignore files under third_party/.
+  skipThirdParty,
+
+  /// Ignore due to a well-formed FLUTTER_NOLINT comment.
+  skipNoLint,
+
+  /// Fail due to a malformed FLUTTER_NOLINT comment.
+  failMalformedNoLint,
+}
+
+/// A compilation command and methods to generate the lint command and job for
+/// it.
+class Command {
+  /// Generate a [Command] from a [Map].
+  Command.fromMap(Map<String, dynamic> map) :
+    directory = io.Directory(map['directory'] as String).absolute,
+    command = map['command'] as String {
+    file = io.File(path.normalize(path.join(
+      directory.path,
+      map['file'] as String,
+    )));
+  }
+
+  /// The working directory of the command.
+  final io.Directory directory;
+
+  /// The compilation command.
+  final String command ;
+
+  /// The file on which the command operates.
+  late final io.File file;
+
+  static final RegExp _pathRegex = RegExp(r'\S*clang/bin/clang');
+  static final RegExp _argRegex = RegExp(r'-MF \S*');
+
+  String? _tidyArgs;
+
+  /// The command line arguments of the command.
+  String get tidyArgs {
+    return _tidyArgs ??= (() {
+      String result = command;
+      result = result.replaceAll(_pathRegex, '');
+      result = result.replaceAll(_argRegex, '');
+      return result;
+    })();
+  }
+
+  String? _tidyPath;
+
+  /// The command but with clang-tidy instead of clang.
+  String get tidyPath {
+    return _tidyPath ??= _pathRegex.stringMatch(command)?.replaceAll(
+      'clang/bin/clang',
+      'clang/bin/clang-tidy',
+    ) ?? '';
+  }
+
+  /// Whether this command operates on any of the files in `queries`.
+  bool containsAny(List<io.File> queries) {
+    return queries.indexWhere(
+      (io.File query) => path.equals(query.path, file.path),
+    ) != -1;
+  }
+
+  static final RegExp _nolintRegex = RegExp(
+    r'//\s*FLUTTER_NOLINT(: $issueUrlPrefix/\d+)?',
+  );
+
+
+  LintAction? _lintAction;
+
+  /// The type of lint that is appropriate for this command.
+  Future<LintAction> get lintAction async =>
+    _lintAction ??= await _getLintAction();
+
+
+  Future<LintAction> _getLintAction() async {
```

Comment:
I'll give this some thought. Available options for testing the functionality are abstracting the input file stream, providing a fake file system implementation, or writing to the /tmp file system, none of which are very attractive.

Reply:
<light hearted rant follows, feel free to ignore it since it isn't a huge deal>

I'm not suggesting writing tests per say, just suggesting that factoring out pure functions (ignoring the filesystem) makes them more testable and by extension more maintainable since they are able to be executed in isolation and the contract of input versus output is more clear than a method (who can touch any instance variable, thus having less encapsulation).

This one was so close to being a static function I thought I'd mention it in my Quixotic quest to make our codebase more Functional.

Reply:
It's a good cause. I shuffled things around a bit. I was gloomy at first since I didn't want to abstract out the file system, but I found a way around it.